### ISSUE TITLE
AWS S3 Sync Issues

### ISSUE THREAD

**Issue by jamesls** [aws/aws-cli#599]
There have been a few issues with respect to the `sync` command, particularly in the case of syncing down from S3 (`s3 -> local`).  I'd like to try to summarize the known issues as well as a few proposals of possible options, and give people the opportunity to share any feedback they might have.
## Sync Behavior Overview

The sync behavior is intended to be an efficient `cp`; only copy over the files from the source to the destination that are different.  In order to do that we need to be able to determine whether or not a file in s3/local are different.  To do this, we use two values:
- File Size (from `stat`'ing the file locally and from the `Size` key in a `ListObjects` response)
- Last modified time (mtime of the local file and the `LastModified` key in a `ListObjects` response)

As an aside, we use the `ListObjects` operation because we get up to 1000 objects returned in a single call.  This means that we're limited to information that comes back from a `ListObjects` response which is `LastModified, ETag, StorageClass, Key, Owner, Size`.

Now given the remote and local files file size and last modified times we try to determine if the file is different.  The file size is easy, if the file sizes are different, then we know the files are different and we need to sync the file.  However, last modified time is more interesting.  While the mtime of the local file is a true mtime, the `LastModified` time from `ListObjects` is really the time the object was uploaded.  So imagine this scenario:

```
aws s3 sync local/ s3://bucket/
sleep 10
aws s3 sync s3://bucket local/
```

After the first sync command (`local->s3`), the local files will have an mtime of 0, and the contents in s3 will have a LastModified time of 10 (using relative offsets).  When we run the second aws s3 sync command, which is syncing from s3 to local we'll first do the file size check.  In this case the file sizes are the same so we look at the last modified time checks.  In this case they are different (local == 0, s3 == 10).  If we were doing a strict equality comparison then, because the last modified times are different, we would unnecessarily sync the files from s3 to local.  So we can say that if the file sizes are the same and the last modified time in s3 is greater (newer) than the local file, then we don't sync.  This is the current behavior.

However, this creates a problem if the remote file is updated out of band (via the console or some other SDK) and the size remains the same.  If we run `aws s3 sync s3://bucket local/` we will not sync the remote file even though we're suppose to.
## Potential Solutions

Below are potential solutions.
1. Change the time checks to be a strict equality comparison.  If the times are different, we sync.  This has the issue that `aws s3 sync local s3://bucket && aws s3 s3://bucket local` will unnecessarily sync files.  However, when we **download** a file, we set the mtime of the file to match the LastModified time so if you were to run `aws s3 sync s3://bucket local` _again_, it would not sync any files.
2. Modify the local mtimes when files are uploaded to match the Last Modified time of S3.  We'd then change the file time checks to a strict equal/not equal check.  The only downside to this approach is whether or not people would expect us to mess with the mtime of their files (I would expect not).
3. Add custom metadata to each object.  This could potentially add some combination of local mtime, md5 so we can also compare md5 checksums if we want.  However, we would have to switch from ListObjects to a HeadObject for every request.  This would require 1000 times more API calls, and would slow down the sync in the case of many small files.  In the steady state case as long as we can HeadObjects faster that we can download/upload them there shouldn't be an issue.  A downside to this approach is that other tools won't be adding this metadata, which would make interop with other tools harder (if an object is missing the metadata we'd likely just sync).
4. Keep a local ETag cache.  We could associate the file name with an ETag/local mtime/last modified time/md5 of the file.  On the plus side we still use ListObjects, so we get 1000 objects back at a time.  Downside is that every client will have to keep a cache.  This is also the most complex solution in terms of implementation.

If there any other potential solutions I've left out, please chime in.

**Comment by Seraf** [comment]
Hello,
I also have this issue well described with foo="bar"/foo="baz".
I use S3 bucket for my application deployment and all servers sync from S3 when a deploy is done. I had few times the problem of an operator >= changed to <= in a file not synced due to this bug and for me sync command is not very reliable due to this bug. The size is the same but the content is different, the file should be synced.
I have no particular advice about how to do it, sorry for that, but I'm just exposing my usecase :)

**Comment by evanplaice** [comment]
Go figure, I came across the same issue while developing node-ftpsync. I assumed AWS would have some magical solution to solve this.

1. Sucks but it's necessary. A mtime.equals comparison is the quickest/dirtiest way to detect change. Unfortunately, the clocks times between the local and remote hosts need to be normalized before they can be compared.

It's probably a good idea to 'fuzz' it (ie rounding to the nearest 10 minutes) like @ngbranitsky suggested. Doing so in node.js is a pain but in python it should be as easy as truncating the last few bits using a bitwise AND.

2. May work, with side-effects. I explored this option also by possibly changing the local mtime to match the server's MDTM (ie FTP mtime) value. Unfortunately, FTP server implementations are rarely RFC-compliant and may not include MDTM. Case in point, file syncing over FTP sucks because there's no guarantee that both the client and server are complete.

Since AWS doesn't have that issue you also have to consider how mtime is used on the local host. By changing the mtime value on every sync are you going to trigger a mass update event if there are transpilers watching those files? Are there other metadata stores that use mtime as a metric for measuring file changes?

3. Is the best option if you don't factor in the cost. Additional API calls incur a cost to the users (ie as in monetary cost). The MD5 will only be necessary for file additions and changed files but for users storing a LOT of small files and/or files that change frequently, it can add up.

4. Is by-far the best option to ensure consistency but also the biggest pain to implement.

**Comment by Frangible** [comment]
I realize this is not a general solution, but it would be very nice to see the following optional behavior for syncing, imo.  It's sort of like no. 4 in the OP, only modified for better performance.

1.  A local cache of etags is kept for local filesystem, and optionally the server if no out-of-band changes to files in the bucket can be guaranteed by the user.
2.  Upon syncing local->server, a first-pass using mtime/size compare against the cache DB is used to determine potential changes in local files.  (This is done to avoid unnecessarily recomputing the md5 which is expensive relative to fstat.)
3.  For the potentially changed set, the md5->etag is recomputed.  If the file is actually different, it is uploaded.  The local cache DB is updated in any case to prevent recomputing the md5 on local files.

The use case here is syncing web map tiles, of which <1% normally change on a daily basis.  The exception is when deletes occur in the source data that change the spatial coverage of the tiles, which necessitates the entire set of map tiles be regenerated.

The issue is the sheer volume involved.  I see discussion about large multipart uploads as a use case, but not about many individual small files.  How many?  We have ~2m currently, but it can get much worse.  For example, at zoom level 16, the world has 1 << 16 x 1 << 16, or 65536 x 65536 tiles, or ~4b.

Current options are:

1.  Sync with aws-cli normally, reuploading everything.
 1. This is the fastest overall.
 2. Alas, many unnecessary uploads will occur.
2.  Sync with s3cmd, which uses md5 (etag) rather than mtimes.  Unfortunately:
 1. It always recomputes the md5 for every local file every time it is run, which is relatively expensive and a lot of IOPS.
 2. It is single threaded and memory-bound, and while always slow, it is especially slow for uploading many individual files, particularly in cases of higher latency to other S3 regions.

I could write C code without much difficulty that would walk a directory path, update a local sqlite3 cache DB, and build the potentially changed set / upload queue.  Unfortunately, I have no python experience and cannot submit this as a pull request for an optional sync behavior.

I don't know how common the "many small files, guaranteed no out-of-band changes, local->server only" use case is.

**Comment by jbarlow83** [comment]
I find it unintuitive and dangerous that sync treats a *difference* in last modified times as a reason to update, even if a newer file at source is replacing an older at the destination. This odd behavior should be documented.

Also related to #404.

**Comment by palesz** [comment]
What about adding a single ```--sync-if``` flag instead of the growing number of options (size-only, exact timestamps, ...)?
The ```--sync-if``` could have a list of options:

```--sync-if newer-timestamp,different-md5,different-timestamp,different-size,deleted,...```

The user could specify one or more (comma separated list) and if the file meets any of the criterias, it will be updated (uploaded/deleted) at the destination.

**Comment by davidmaxwaterman** [comment]
Reading through this issue, I can't figure out if this sync behaviour has been fixed yet.

I want something that works as simply as `rsync -avz` to sync my local build with those on the server.

I had been using `aws s3 sync`, but because I have one file which is large (a help file that is a movie) and my build step creates all files from new, it copies all files every time and was needlessly slow to update the site.

I then started using `--size-only` to speed it up. Unfortunately, this bit me in the butt recently. I had renamed a file, and the build step includes a list of files in the service-worker.js file so the sw knows what to cache. Unfortunately, the new and old file names were the same length, so it didn't update the service-worker.js file, and kept giving a 404 for the old filename. It took me quite a while to figure out what was going on.

It does seem like this is a solved problem in other environments - ie by rsync - though, tbh, I'm probably somewhat ignorant of the challenges presented with doing something like this on S3. Anyway, having being bit by this recently, I'm looking at other clients, but they seem to have dependencies that I'm not interested in adopting just for this functionality.

**Comment by simlu** [comment]
Would be great to have an etag sync option. I know that there are scenarios where it fails, but for me it would be super valuable.

**Comment by ASayre** [comment]
Good Morning!

We're closing this issue here on GitHub, as part of our migration to [UserVoice](https://aws.uservoice.com/forums/598381-aws-command-line-interface) for feature requests involving the AWS CLI.

This will let us get the most important features to you, by making it easier to search for and show support for the features you care the most about, without diluting the conversation with bug reports.

As a quick UserVoice primer (if not already familiar): after an idea is posted, people can vote on the ideas, and the product team will be responding directly to the most popular suggestions.

We’ve imported existing feature requests from GitHub - Search for this issue there!

And don't worry, this issue will still exist on GitHub for posterity's sake.  As it’s a text-only import of the original post into UserVoice, we’ll still be keeping in mind the comments and discussion that already exist here on the GitHub issue.

GitHub will remain the channel for reporting bugs.  

Once again, this issue can now be found by searching for the title on: https://aws.uservoice.com/forums/598381-aws-command-line-interface 

-The AWS SDKs & Tools Team

**Comment by ITmaze** [comment]
I am unable to locate this issue on UserVoice. Perhaps I'm looking with my ears, but a link would be appreciated.

**Comment by four43** [comment]
Moving away from github issues? That seems like a mistake...

**Comment by davidmaxwaterman** [comment]
Agreed. It seems more like the method Microsoft use for judging the importance/impact of issues, but I find it quite irritating.

**Comment by jamesls** [comment]
Based on community feedback, we have decided to return feature requests to GitHub issues.

**Comment by acejam** [comment]
Bumping this back up!

**Comment by acejam** [comment]
To the top!

**Comment by queglay** [comment]
comparison of md5 would be great, I'd also add that it would be helpful to output md5 on upload or download.  this could be stored in our own db and would help determining if sync is needed through our database to limit requests.

**Comment by danil-smirnov** [comment]
@jamesls Could you comment this issue please?
https://github.com/aws/aws-cli/issues/4460

**Comment by BourgoisMickael** [comment]
Any news about this after **8 years** ? Will the solution be implemented before the end of the century ?

**Comment by Zenexer** [comment]
@BourgoisMickael It was already implemented: `--exact-timestamps`

**Comment by muliyul** [comment]
> @BourgoisMickael It was already implemented: `--exact-timestamps`

This is relevant to S3 -> local sync ops but not the other way around (if I understand correctly).

**Comment by chambois** [comment]
Any update on this? Found this issue from #9074 and it seems quiet here. 

We were caught very unpleasantly by surprise by `aws s3 sync` command - we assumed (apparently falsely) that by providing the `--exact-timestamps` flag, we would be guaranteed that all the changed files would be replaced. This, however, was not the case when we rolled back and our `index.html` was left (and not replaced as it should have been). My guess is because we were trying to replace it with an older version, `sync` left the latest one there - which is wrong in this case and not the expected behaviour.

Can you please:

- Update the docs (`aws s3 sync help`) to be much clearer about how exactly `--exact-timestamps` behaves, so that other people aren't left dealing with unnecessary production issues like we were
- Implement a robust option for syncing based on checksum (as many others have requested) 
- Provide us all an ETA on when we can expect these changes

Thanks!

**Comment by chambois** [comment]
Also, in the meantime - can you recommend how we can best deal with the case when we want to sync files with the same filename and size, but an older timestamp?

Thanks!

**Comment by asoloa** [comment]
To those with the same issue as me, I hope this helps.

In my GitHub Actions workflow using GitHub-hosted runner, my steps include checking out the latest source code, then syncing it to s3.
Since the timestamps of the checked out files will always be the time that they were checked out, `aws s3 sync` will always upload even with no changes.

I worked around this issue by utilizing Ansible's [s3_sync module](https://docs.ansible.com/ansible/latest/collections/community/aws/s3_sync_module.html) which supports "checksum" for its [file_change_strategy](https://docs.ansible.com/ansible/latest/collections/community/aws/s3_sync_module.html#parameter-file_change_strategy) parameter. Executing the task via ad-hoc command:
```bash
ansible localhost -c local -e "ansible_python_interpreter=/usr/bin/python3.12" -m community.aws.s3_sync \
-a "bucket=${{ BUCKET }} file_root='.' file_change_strategy='checksum' delete=true" | tee /tmp/ansible.log
```
Also note that Ansible is already installed in [GitHub-hosted runner image](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md) that I've used.

**Comment by pete-woods** [comment]
11 years now. Please implement support for this - I don't want to have to write my own sync tool

**Comment by cmdr2** [comment]
In case this helps anyone, here's a script that syncs only the changed files - https://github.com/cmdr2/s3-sync-changes

Example:
`python s3-sync-changes.py . s3://my-bucket/path/to/dir --exclude .git --exclude README.md --acl public-read`

### How
It checks the Etags (by calling ListBucket and comparing the Etags of the local files). Probably not the most elegant way, but works for small-sized buckets (< 10000 objects). And it protects against over-calling ListBucket with a default `--max-objects` of 3000.

By default, it'll sync only 3000 objects, but that limit can be increased with the `--max-objects N` argument.

### Why
I needed it for syncing blogs and websites using GitHub Actions. Much to my shock and surprise, I discovered (after many months) that 's3 sync' was uploading *all* of my content each time (every image and gif!).

PS: Please keep in mind that List requests cost more, so this script isn't meant for large buckets or buckets with frequent syncs.

**Comment by RyanFitzSimmonsAK** [comment]
Thank you all for your interest in this issue.  This issue has served to track many bugs and feature requests about the sync command, but it's reached a point where individual issues are hard to follow here.  We're going to break this down into separate issues to be tracked and addressed individually.  I've identified the following bugs and feature requests mentioned here, and if I've missed any, please feel free to let me know and I'll make another issue to track it.

- Checksum comparison : There's several duplicates of this particular feature request; https://github.com/aws/aws-cli/issues/9074 will be reopened and promoted to the tracking issue.

- `--update` flag similar to rsync :  https://github.com/aws/aws-cli/issues/404 tracks this.

- apply `--exact-timestamps` flag to uploads : https://github.com/aws/aws-cli/issues/4460 tracks this.

- fuzzy time comparisons : I've created a new issue, https://github.com/aws/aws-cli/issues/10017, to track this.  This feature request will go through our standard triage and review process.
	
- sync-if enum : I've created a new issue, https://github.com/aws/aws-cli/issues/10016, to track this.  This feature request will go through our standard triage and review process.
	
Additionally, the `--no-overwrite` parameter was recently added to high-level AWS S3 operations (AWS CLI version `2.33.0`), including ``sync``.  This parameter prevents overwriting objects that already exist at the destination, which you might find useful for some of the use cases mentioned in this issue.
